### PR TITLE
Fixes allowing loading HF Llama checkpoints and saving torch_dist checkpoints

### PR DESCRIPTION
**Pull request by aivanni**
1. Using `--no-gradient-accumulation-fusion`  flag during llama huggingface checkpoint loading, due to APEX not supported.

2. Adjust checkpoints saving in torch_dist format to account for recent breaking change (https://github.com/pytorch/pytorch/pull/143359) in torch distributed library that is e.g. in rocm/pytorch-training:v25.2 image already.

Related PR #62

### CODE REVIEW

**Review by sarooshsh — APPROVED** [review]

File: megatron/training/checkpointing.py [hunk 1]
```diff
@@ -48,6 +48,13 @@
 except Exception:
     has_nvidia_modelopt = False
 
+# Imports needed to allow loading checkpoints previously saved with saver_mcore
+from argparse import Namespace
+from megatron.core.enums import ModelType
+from io import BytesIO
+
+torch.serialization.add_safe_globals([Namespace, ModelType, BytesIO])
+
```
Comment (mpashkovskii) [inline megatron/training/checkpointing.py]: > 2. Allow loading checkpoints previously converted using llama_mistral loader and mcore saver. Without it you'll see errors like:
>
> ```
> could not load the checkpoint
> Weights only load failed. This file can still be loaded, to do so you have two options, do those steps only if you trust the source of the checkpoint. 
> 	(1) Re-running `torch.load` with `weights_only` set to `False` will likely succeed, but it can result in arbitrary code execution. Do it only if you got the file from a trusted source.
> 	(2) Alternatively, to load with `weights_only=True` please check the recommended steps in the following error message.
> 	WeightsUnpickler error: Unsupported global: GLOBAL argparse.Namespace was not an allowed global by default. Please use `torch.serialization.add_safe_globals([Namespace])` or the `torch.serialization.safe_globals([Namespace])` context manager to allowlist this global if you trust this class/function.
> 
> Check the documentation of torch.load to learn more about types accepted by default with weights_only https://pytorch.org/docs/stable/generated/torch.load.html.
> ```

I'm not sure that it the correct fix. I think that's the correct one https://github.com/ROCm/Megatron-LM/pull/56

Comment (aivanni) [reply megatron/training/checkpointing.py]: Thanks for pointing to that, missed it, will re-test with latest commit

Comment (aivanni) [reply megatron/training/checkpointing.py]: I tested checkpoint loading with the latest commit and that worked well, so reverting these lines

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -469,10 +469,6 @@ def __init__(
         for model_chunk in self.model_chunks:
             assert self.ddp_config == model_chunk.ddp_config
 
-        assert isinstance(
```
Comment (wenchenvincent) [inline megatron/core/optimizer/distrib_optimizer.py]: Please don't remove that as we don't have evidence.

**Review by wenchenvincent — CHANGES_REQUESTED** [review]

File: megatron/core/optimizer/distrib_optimizer.py (hunk 2 above)
Comment (aivanni) [reply megatron/core/optimizer/distrib_optimizer.py]: I tested code with sgd optimizer and it worked well (full cycle, checkpoint loading, training, checkpoint saving). can revert that sure, but can also run additional tests if you'd like to add some confirmations. if you can suggest what cases should be verified?

Comment (wenchenvincent) [reply megatron/core/optimizer/distrib_optimizer.py]: Notice that NV upstream adds the support for HybridDeviceOptimizer but still does not support SGD (https://github.com/NVIDIA/Megatron-LM/blob/ed7db1d2b3e03378c88d50707b181b485882b030/megatron/core/optimizer/distrib_optimizer.py#L494-L497). I believe there is some reason that we don't fully understand regarding why the distributed optimizer does not work for SGD. Let's revert that.

Comment (aivanni) [reply megatron/core/optimizer/distrib_optimizer.py]: Reverted

**Comment by aivanni** [comment]
@wenchenvincent some tests are failing. trying to check the details and currently jenkins site is not accessible. But when it was working it was asking for credentials, which ones should we use there? Shall we apply for the access somehow? Also why do tests take so long? Do tests run GPU benchmarks and most of this time is waiting for spare GPU resources?

**Comment by aivanni** [comment]
@wenchenvincent @gurpreet-dhami do you have other comments that should be addressed?

File: megatron/core/optimizer/distrib_optimizer.py (hunk 2 above)
Comment (aivanni) [reply megatron/core/optimizer/distrib_optimizer.py]: @wenchenvincent in the llama training examples we have support for `sgd` https://github.com/ROCm/Megatron-LM/blob/d6dfaede35896420e55c960403f9cc173ac5c3ac/examples/llama/train_llama3.sh#L210

File: megatron/core/dist_checkpointing/strategies/filesystem_async.py [hunk 3]
```diff
@@ -20,6 +20,15 @@
 from torch.distributed.checkpoint.storage import WriteResult
 from torch.futures import Future
 
+try:
+    # This PR https://github.com/pytorch/pytorch/pull/143359 introduced breaking change to saving checkpoints
```
Comment (wenchenvincent) [inline megatron/core/dist_checkpointing/strategies/filesystem_async.py:24]: In which pytorch version it would break?

File: megatron/core/optimizer/distrib_optimizer.py (hunk 2 above)
Comment (vidushi8) [reply megatron/core/optimizer/distrib_optimizer.py]: By default we keep as adam in training scripts.

**Review by vidushi8 — COMMENTED** [review]
LGTM.

**Review by vidushi8 — APPROVED** [review]

File: megatron/core/dist_checkpointing/strategies/filesystem_async.py (hunk 3 above)
Comment (aivanni) [reply megatron/core/dist_checkpointing/strategies/filesystem_async.py:24]: @wenchenvincent This breaks for example with torch version `2.7.0a0+git6374332` that is used in `rocm/megatron-lm:v25.4`

**Review by wenchenvincent — APPROVED** [review]
